### PR TITLE
Feature/#12 diff functionality

### PR DESCRIPTION
**Pull request by TheGreatRefrigerator**
Adds boolean `generate_difference` to `portal_options` for the ors_api POST endpoints.

Passing `true` consumes 2 ors requests and internally sends one request with the stored avoid areas and one without.
The geometric difference is calculated, relevant metrics recaclulated and irrelevant attributes dropped.
The result is returned in either `geojson` format or `json` for directions only. `gpx` is not supported.

Closes https://github.com/GIScience/heigit-disaster-portal/issues/12

### CODE REVIEW

File: dap_api/app/app/backend/geoutil.py [hunk 1]
```diff
@@ -110,3 +113,45 @@ def meters_travelled(seconds, speed):
...
+    @param ors_res_type: response format
+    @return: sqlalchemy query function
+    """
+    geom_no_avoid = get_geom_from_item(item, ors_res_type)
+    geom = get_geom_from_item(avoid_item, ors_res_type)
+    difference = None
+    if ors_api == "isochrones":
+        difference = func.ST_Difference(geom_no_avoid, geom)
+    elif ors_api == "directions":
+        difference = func.ST_Difference(geom, geom_no_avoid)
```
Comment (koebi) [inline dap_api/app/app/backend/geoutil.py:135]: This is confusing enough that I'd recommend a comment explaining why it has to be the other way round for directions.

Comment (TheGreatRefrigerator) [reply dap_api/app/app/backend/geoutil.py:135]: agreed

File: dap_api/app/app/backend/ors_processor.py [hunk 2]
```diff
@@ -52,15 +53,29 @@ def handle_ors_request(self, db: Session, request: ORSRequest, options: PathOpti
         # relay to backend
         endpoint = f"/{options.ors_api}/{options.ors_profile}/{options.ors_response_type}"
         response = self.relay_request_post(endpoint, request_header, request_dict)
+        response_json = {}
+        if response.status_code == 200 and request.portal_options.generate_difference:
+            if request_dict.get('options'):
+                request_dict.get('options').pop('avoid_polygons')
```
[the remark this answers is not in the document]
Comment (TheGreatRefrigerator) [reply dap_api/app/app/backend/ors_processor.py:59]: good point! done

File: dap_api/app/app/backend/ors_processor.py [hunk 3]
```diff
@@ -52,15 +53,29 @@ def handle_ors_request(self, db: Session, request: ORSRequest, options: PathOpti
...
+            response_no_avoid = self.relay_request_post(endpoint, request_header, request_dict)
+
+            response_json = response.json()
+            new_features = self.calculate_new_features(db, options, request_dict, response.json(),
```
[the remark this answers is not in the document]
Comment (TheGreatRefrigerator) [reply dap_api/app/app/backend/ors_processor.py:63]: also correct. done

File: dap_api/app/app/backend/ors_processor.py [hunk 4]
```diff
@@ -69,6 +84,58 @@ def handle_ors_request(self, db: Session, request: ORSRequest, options: PathOpti
...
+            avoid_item['geometry'] = json.loads(diff_geom)
+            # no difference
+            if not avoid_item['geometry']['coordinates']:
+                continue
+            ORSProcessor.update_info(avoid_item, item, options, request_dict)
+            new_features.append(avoid_item)
+        return new_features
+
+    @staticmethod
+    def get_next_avoid_item(avoid_results, item, options):
```
[the remark this answers is not in the document]
Comment (TheGreatRefrigerator) [reply dap_api/app/app/backend/ors_processor.py:109]: I've changed to iterating with the index in the loop outside this function and change this function to extract the matching item only for isochrones

**Pull request merged**